File: source/python/simple_chat.py
import threading
import queue

alice_q = queue.Queue(maxsize=10)
bob_q = queue.Queue(maxsize=10)

def alice(send_q, recv_q):
    send_q.put("Hi Bob! How are you?")
    msg = recv_q.get()
    print(f"Alice received: {msg}")

    send_q.put("I'm working on a Hemlock project!")
    msg = recv_q.get()
    print(f"Alice received: {msg}")

    send_q.put("Thanks! Talk later.")
    msg = recv_q.get()
    print(f"Alice received: {msg}")

def bob(send_q, recv_q):
    msg = recv_q.get()
    print(f"Bob received: {msg}")
    send_q.put("Hey Alice! I'm great, you?")

    msg = recv_q.get()
    print(f"Bob received: {msg}")
    send_q.put("That sounds awesome!")

    msg = recv_q.get()
    print(f"Bob received: {msg}")
    send_q.put("Bye Alice!")

a = threading.Thread(target=alice, args=(bob_q, alice_q))
b = threading.Thread(target=bob, args=(alice_q, bob_q))

a.start()
b.start()
a.join()
b.join()

print("Chat ended.")

File: source/rust/simple_chat.rs
use std::sync::mpsc;
use std::thread;

fn main() {
    let (alice_tx, alice_rx) = mpsc::channel::<String>();
    let (bob_tx, bob_rx) = mpsc::channel::<String>();

    let alice_tx2 = alice_tx.clone();
    let bob_tx2 = bob_tx.clone();

    let alice = thread::spawn(move || {
        bob_tx.send("Hi Bob! How are you?".to_string()).unwrap();
        let msg = alice_rx.recv().unwrap();
        println!("Alice received: {}", msg);

        bob_tx.send("I'm working on a Hemlock project!".to_string()).unwrap();
        let msg = alice_rx.recv().unwrap();
        println!("Alice received: {}", msg);

        bob_tx.send("Thanks! Talk later.".to_string()).unwrap();
        let msg = alice_rx.recv().unwrap();
        println!("Alice received: {}", msg);
    });

    let bob = thread::spawn(move || {
        let msg = bob_rx.recv().unwrap();
        println!("Bob received: {}", msg);
        alice_tx2.send("Hey Alice! I'm great, you?".to_string()).unwrap();

        let msg = bob_rx.recv().unwrap();
        println!("Bob received: {}", msg);
        alice_tx2.send("That sounds awesome!".to_string()).unwrap();

        let msg = bob_rx.recv().unwrap();
        println!("Bob received: {}", msg);
        alice_tx2.send("Bye Alice!".to_string()).unwrap();
    });

    alice.join().unwrap();
    bob.join().unwrap();
    println!("Chat ended.");
}
